pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -98,11 +145,64 @@ type DockercfgController struct {
 
 	serviceAccountCache      MutationCache
 	serviceAccountController *framework.Controller
+	secretsIndex             cache.Indexer
+	secretController         *framework.Controller
 
 	queue workqueue.RateLimitingInterface
 
 	// syncHandler does the work. It's factored out for unit testing
 	syncHandler func(serviceKey string) error
+
+	// record the token secret we created to pendingTokens and prevent creating a new tokens
+	// while the cache is populated
+	pendingTokens *pendingTokenStore
+}
+
+// handleTokenSecretUpdate checks if the service account token secret is populated with
+// token data and triggers re-sync of service account when the data are observed.
+func (e *DockercfgController) handleTokenSecretUpdate(_, newObj interface{}) {
+	secret := newObj.(*api.Secret)
+	serviceAccountName, _ := secret.Annotations[api.ServiceAccountNameKey]
+	serviceAccountKey := secret.Namespace + "/" + serviceAccountName
+	obj, exists, err := e.serviceAccountCache.GetByKey(serviceAccountKey)
+	if err != nil {
+		glog.V(2).Infof("Unable to retrieve the service account %q from cache: %v", serviceAccountKey, err)
+		return
+	}
+	if !exists {
+		return
+	}
+	// The secret is now available in the cache so remove it from pendingTokens
+	e.pendingTokens.Delete(serviceAccountKey)
+	if createdBy, exists := secret.Annotations[api.CreatedByAnnotation]; exists && createdBy == CreateDockercfgSecretsController {
+		if len(secret.Data[api.ServiceAccountTokenKey]) > 0 {
+			e.enqueueServiceAccount(obj.(*api.ServiceAccount))
+		}
+	}
+}
+
+// handleTokenSecretDelete handles token secrets deletion and re-sync the service account
+// which will cause a token to be re-created.
+func (e *DockercfgController) handleTokenSecretDelete(obj interface{}) {
+	secret, isSecret := obj.(*api.Secret)
+	if !isSecret {
+		tombstone, objIsTombstone := obj.(cache.DeletedFinalStateUnknown)
+		if !objIsTombstone {
+			glog.V(2).Infof("Expected tombstone object when deleting token, got %v", obj)
+			return
+		}
+		secret, isSecret = tombstone.Obj.(*api.Secret)
+		if !isSecret {
+			glog.V(2).Infof("Expected tombstone object to contain secret, got: %v", obj)
+			return
+		}
+	}
+	serviceAccountName, _ := secret.Annotations[api.ServiceAccountNameKey]
+	if secret.Annotations[api.CreatedByAnnotation] == CreateDockercfgSecretsController {
+		if obj, exists, _ := e.serviceAccountCache.GetByKey(secret.Namespace + "/" + serviceAccountName); exists {
+			e.enqueueServiceAccount(obj.(*api.ServiceAccount))
```

Comment:
@liggitt so you want to check if the token data is set?  (iow. re-sync just for population pending tokens?)